docs/hld/chel-fs-hld.md

```diff
@@ -0,0 +1,162 @@
+1. DAOS Containers (Extracts from https://docs.daos.io/v2.6/overview/storage/)
+   1. "A container represents an object address space inside a DAOS pool"
+      This means all the objects of a Chel-FS filesystem can be tucked in a container, like a unique namespace.
+   2. Each container has it's own attributes and capabilities & once opened an application "may share this handle with any or all of its peers" processes
+      This means all the MDSs which are serving a Chel-FS filesystem will be sharing a handle with all it's capabilities and authorizations.
+   3. "Objects in a container may have different schemas for data distribution and redundancy over targets"
+      This means a Chel-FS filesystems can map it's metadata object to a different distribution and redundancy scheme than the data objects.
+      Metadata Object can use 3 way replica and Data Object can use Erasure Coding.
+   4. "A DAOS object can be accessed through different APIs
+        Multi-level key-array, Key-value and Array API"
+        (https://github.com/daos-stack/daos/blob/master/src/object/README.md)
+      This means Chel-FS entities like Directory and Files can be easily mapped to one of these Object Types.
+      (inspiration from https://github.com/daos-stack/daos/blob/master/src/client/dfs/README.md)
+   5. "A container is the basic unit of transaction and versioning" & "The DAOS transaction API allows to combine multiple object updates into a single atomic transaction"
+      This means MDSs can use this transaction API for multiple entities atomic updates
+      Also a container snapshot would be mapped to a Chel-FS snapshot.(more about this [later](#chel-fs-snapshots))
+
+### Distributed Transaction Synchronization
+DAOS container does provide transaction which involves 1 or many Objects (which can be any of the filesystem entities).
+As mentioned in https://docs.daos.io/v2.6/overview/transaction/#distributed-transactions
+```
+Unlike POSIX, the DAOS API does not impose any worst-case concurrency control mechanism to address  
+conflicting I/O operations. Instead, individual I/O operations are tagged with a different epoch  
+and applied in epoch order, regardless of execution order. This baseline model delivers the maximum  
+scalability and performance to data models and applications that do not generate conflicting I/O workload.
+```
+This is incredible but for MDS Transaction (metadata operation which could involve multiple entities) would required a level of isolation when the transactions happen. MDS Sharding model makes sure that no 2 MDS would work on same entities for most of the metadata operations. But some metadata operations require some level of coordination between 2 MDS eg: Rename/Move of file/Directories between 2 Parent directories. In this case 1 MDS would signal the other MDS that it would do the transaction on behalf of both, i.e unlink on 1st Parent and link on the 2nd Parent.
```

Comment:
Why can't you reply on transaction for coordination between 2 MDs? The reason is not clear here.

Reply:
I think you are correct. Transaction coordination may not be required between MDS1 and MDS2. I hope in scenarios where 
Scenarios 1 : 

1. If object 1 is (Logically)owned by MDS1 and if Object 2 is (Logically) owned by MDS 2. 
2.  MDS1 changes Object 1 & Object 2 in a transaction
3. And MDS2 deletes the Object2 in another transaction,
I hope the correctness is maintained in the order of the transaction epoc.

Also

Scenario 2:
1. If object 1 is (Logically)owned by MDS1 and if Object 2 is (Logically) owned by MDS 2. 
2.  MDS1 changes Object 1 & Object 2 in a transaction

Do we have callbacks on change of objects in DAOS i.e willl MDS 2 will be notified by DAOS on the change of Object 2, Do we have a subscribe+notify mechasim in DAOS
If yes then we can avoid doing a RPC Call between MDS1 and MDS2 to notify MDS2
If no then we need a notification mechanism to inform MDS2

Reply:
There is only one transaction can success if two transaction are modifying the object. The other one needs to refresh and retry.

Reply:
I dont think it does compare and exchange. (failing an atomic exchange when a condition fails)

Could you please elaborate with the example scenarios ?

Reply:
It is similar to compare and swap.  It is MVCC. You operate on a snapshot of object states.  But if the dependent states have changed,  the transaction needs to start all over.

Reply:
For example one user wants to move file1 to dir2,  and another user wants to rename file1 to dir3. These two transactions will depend on state that there is a dkey file1 in its parent directory. Only one transaction will succeed. The other transaction will fail because the object key existence state changed.